### PR TITLE
Feat/switch to 2018.3 prefabs

### PR DESCRIPTION
**Pull request by thestonefox**

### CODE REVIEW

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs [hunk 1]
```diff
@@ -1,64 +1,69 @@
...
+    using VRTK.Core.Data.Attribute;
+
+    /// <summary>
+    /// Sets up the Interactable.Climbable prefab based on the provided user settings.
+    /// </summary>
+    public class ClimbInteractableInternalSetup : MonoBehaviour
+    {
+        #region Facade Settings
+        [Header("Facade Settings"), Tooltip("The public interface facade."), InternalSetting, SerializeField]
+        private ClimbInteractableFacade facade = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs [hunk 2]
```diff
@@ -1,64 +1,69 @@
...
+        #endregion
+
+        #region Reference Settings
+        /// <summary>
+        /// The <see cref="InteractableFacade"/> component acting as the interactable for climbing.
+        /// </summary>
+        [Header("Reference Settings"), Tooltip("The Interactable Facade component acting as the interactable for climbing."), InternalSetting, SerializeField]
+        private InteractableFacade interactableFacade = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs [hunk 3]
```diff
@@ -1,64 +1,69 @@
...
+        /// <summary>
+        /// The <see cref="GameObjectEventProxyEmitter"/> component handling a started climb.
+        /// </summary>
+        [Tooltip("The Game Object Event Proxy Emitter component handling a started climb."), InternalSetting, SerializeField]
+        private GameObjectEventProxyEmitter startEventProxyEmitter = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs [hunk 4]
```diff
@@ -1,64 +1,69 @@
...
+        /// <summary>
+        /// The <see cref="GameObjectEventProxyEmitter"/> component handling a stopped climb.
+        /// </summary>
+        [Tooltip("The Game Object Event Proxy Emitter component handling a stopped climb."), InternalSetting, SerializeField]
+        private GameObjectEventProxyEmitter stopEventProxyEmitter = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Climb/ClimbInteractableInternalSetup.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/ClimbInteractableFacade.cs [hunk 5]
```diff
@@ -3,18 +3,20 @@
...
+    using VRTK.Core.Prefabs.Interactions.Interactables.Climb;
 
     /// <summary>
     /// The public interface for the Interactable.Climbable prefab.
     /// </summary>
     public class ClimbInteractableFacade : MonoBehaviour
     {
         #region Climb Settings
+        [Header("Climb Settings"), Tooltip("The Climb Facade to use."), SerializeField]
+        private ClimbFacade _climbFacade = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/ClimbInteractableFacade.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableAction.cs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    using VRTK.Core.Data.Attribute;
+
+    /// <summary>
+    /// Describes an action to perform when a Grab Process is executed.
+    /// </summary>
+    public class GrabInteractableAction : MonoBehaviour
+    {
+        #region Input Settings
+        [Header("Input Settings"), Tooltip("The input ActiveCollisionConsumerEventProxyEmitter for the grab action."), InternalSetting, SerializeField]
+        private ActiveCollisionConsumerEventProxyEmitter _inputActiveCollisionConsumer = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableAction.cs]: Remove ` = null`. **Also do this in the rest of this file.**

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableControlDirectionAction.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        /// <summary>
+        /// A <see cref="GameObject"/> collection to enable/disabled as part of the direction control process.
+        /// </summary>
+        [Header("Interactable Settings"), Tooltip("A GameObject collection to enable/disabled as part of the direction control process."), SerializeField]
+        private List<GameObject> linkedObjects = new List<GameObject>();
+        /// <summary>
+        /// The <see cref="DirectionModifier"/> to process the direction control.
+        /// </summary>
+        [Tooltip("The DirectionModifier to process the direction control."), InternalSetting, SerializeField]
+        private DirectionModifier directionModifier = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableControlDirectionAction.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableFollowAction.cs [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            }
+        }
+        #endregion
+
+        #region Tracking Settings
+        /// <summary>
+        /// The <see cref="ObjectFollower"/> for tracking movement.
+        /// </summary>
+        [Header("Tracking Settings"), Tooltip("The ObjectFollower for tracking movement."), InternalSetting, SerializeField]
+        private ObjectFollower objectFollower = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableFollowAction.cs]: Remove ` = null`. **Also do this in the rest of the file.**

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableScaleAction.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    /// Describes an action that allows the Interactable to be scaled in size between the points of two specified Interactors.
+    /// </summary>
+    public class GrabInteractableScaleAction : GrabInteractableAction
+    {
+        #region Interactable Settings
+        /// <summary>
+        /// The <see cref="PinchScaler"/> to process the scale control.
+        /// </summary>
+        [Header("Interactable Settings"), Tooltip("The PinchScaler to process the scale control."), InternalSetting, SerializeField]
+        private PinchScaler pinchScaler = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Action/GrabInteractableScaleAction.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/GrabInteractableInternalSetup.cs [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+    using VRTK.Core.Prefabs.Interactions.Interactables.Grab.Provider;
+
+    /// <summary>
+    /// Sets up the Interactable Prefab grab settings based on the provided user settings.
+    /// </summary>
+    public class GrabInteractableInternalSetup : MonoBehaviour
+    {
+        #region Facade Settings
+        [Header("Facade Settings"), Tooltip("The public interface facade."), SerializeField]
+        private InteractableFacade _facade = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/GrabInteractableInternalSetup.cs]: Remove ` = null`. **Also do this in the rest of the file.**

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Provider/GrabInteractableInteractorProvider.cs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    using VRTK.Core.Extension;
+
+    /// <summary>
+    /// Processes a received grab event and passes it over to the appropriate grab actions.
+    /// </summary>
+    public abstract class GrabInteractableInteractorProvider : MonoBehaviour
+    {
+        #region Input Settings
+        [Header("Input Settings"), Tooltip("The input GameObjectEventProxyEmitter for the grab action."), InternalSetting, SerializeField]
+        private GameObjectEventProxyEmitter _inputGrabReceived = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Provider/GrabInteractableInteractorProvider.cs]: Remove ` = null`. **Also do this in the rest of the file.**

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Provider/GrabInteractableSetInteractorProvider.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+    using VRTK.Core.Prefabs.Interactions.Interactors;
+
+    /// <summary>
+    /// Processes a received grab event into an Observable Set to handle a simplified grab process.
+    /// </summary>
+    public class GrabInteractableSetInteractorProvider : GrabInteractableInteractorProvider
+    {
+        #region Stack Settings
+        [Header("Set Settings"), Tooltip("The set to get the current interactors from."), InternalSetting, SerializeField]
+        private GameObjectObservableSet _eventSet = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Provider/GrabInteractableSetInteractorProvider.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Provider/GrabInteractableStackInteractorProvider.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    using VRTK.Core.Prefabs.Interactions.Interactors;
+
+    /// <summary>
+    /// Processes a received grab event into an Observable Stack to handle multiple output options for each grab type.
+    /// </summary>
+    public class GrabInteractableStackInteractorProvider : GrabInteractableInteractorProvider
+    {
+        #region Stack Settings
+        [Header("Stack Settings"), Tooltip("The stack to get the current interactors from."), InternalSetting, SerializeField]
+        private GameObjectObservableStack _eventStack = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Provider/GrabInteractableStackInteractorProvider.cs]: Remove ` = null`.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Receiver/GrabInteractableReceiver.cs [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+            }
+        }
+        #endregion
+
+        #region Grab Consumer Settings
+        /// <summary>
+        /// The <see cref="ActiveCollisionConsumer"/> that listens for the grab payload.
+        /// </summary>
+        [Header("Grab Consumer Settings"), Tooltip("The ActiveCollisionConsumer that listens for the grab payload."), InternalSetting, SerializeField]
+        private ActiveCollisionConsumer grabConsumer = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/Receiver/GrabInteractableReceiver.cs]: Remove ` = null`. **Also do this in the rest of the file.**

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/InteractableFacade.cs [hunk 15]
```diff
@@ -141,17 +64,68 @@ public enum GrabOffset
...
+            {
+                _consumerRigidbody = value;
+                ConfigureContainer();
+            }
+        }
+        #endregion
+
         #region Internal Settings
+        [Header("Internal Settings"), Tooltip("The linked CollisionNotifier."), SerializeField, InternalSetting]
+        private CollisionNotifier _collisionNotifier = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/InteractableFacade.cs]: Remove ` = null`. **Also do this in the rest of the file.**

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Touch/TouchInteractableInternalSetup.cs [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    using VRTK.Core.Tracking.Collision;
+    using VRTK.Core.Tracking.Collision.Active;
+    using VRTK.Core.Tracking.Collision.Active.Operation;
+    using VRTK.Core.Prefabs.Interactions.Interactors;
+
+    public class TouchInteractableInternalSetup : MonoBehaviour
+    {
+        #region Facade Settings
+        [Header("Facade Settings"), Tooltip("The public interface facade."), SerializeField]
+        private InteractableFacade _facade = null;
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Touch/TouchInteractableInternalSetup.cs]: Remove ` = null`. **Also do this in the rest of the file.**

File: Scripts/Event/ActiveCollisionsContainerEventProxyEmitter.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
+﻿namespace VRTK.Core.Event
+{
+    using UnityEngine.Events;
+    using System;
+    using VRTK.Core.Tracking.Collision.Active;
+
+    /// <summary>
+    /// Emits a UnityEvent with a <see cref="ActiveCollisionsContainer.EventData"/> payload whenever <see cref="Receive"/> is called.
```
Comment (bddckr) [inline Scripts/Event/ActiveCollisionsContainerEventProxyEmitter.cs]:
```suggestion
    /// Emits a UnityEvent with a <see cref="ActiveCollisionsContainer.EventData"/> payload whenever <see cref="SingleEventProxyEmitter{TValue,TEvent}.Receive"/> is called.
```
According to my Visual Studio only this works.

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/GrabInteractableInternalSetup.cs.meta [hunk 18]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: dd6d3bd046e0d7043aaed76be77f521b
+guid: 8314c9a77122ba744ac0a55af02c58b4
```
Comment (bddckr) [inline Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/GrabInteractableInternalSetup.cs.meta]: Was this on purpose? This GUID change is probably making existing usages in Unity assets break this link to script?

File: Prefabs/CameraRig/SimulatedCameraRig/SharedResources/Scripts/TransformPropertyResetter.cs [hunk 19]
```diff
@@ -13,9 +13,21 @@ public class TransformPropertyResetter : MonoBehaviour
...
+        /// <summary>
+        /// The initial local rotation of the <see cref="source"/>.
+        /// </summary>
         protected Quaternion initialLocalRotation;
+        /// <summary>
+        /// The initial local scale of the <see cref="source"/>.
+        /// </summary>
         protected Vector3 initialLocalScale;
+        /// <summary>
+        /// Whether the initial statews have been set.
```
Comment (bddckr) [inline Prefabs/CameraRig/SimulatedCameraRig/SharedResources/Scripts/TransformPropertyResetter.cs]:
```suggestion
        /// Whether the initial states have been set.
```

File: Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/GrabInteractableInternalSetup.cs.meta (hunk 18 above)
Comment (thestonefox) [reply Prefabs/Interactions/Interactables/SharedResources/Scripts/Grab/GrabInteractableInternalSetup.cs.meta]: The break was caused by a rebase where the script moved locations between the commits (and was actually a git move so it just did a rename)

but it seems that Unity got itself all confused and thought the guid had changed. Fixing this is a reflog job so it's probably best to just push ahead and ignore this ever happened!

File: Prefabs/Locomotion/BodyRepresentation/SharedResources/Scripts/BodyRepresentationInternalSetup.cs [hunk 20]
```diff
@@ -277,13 +289,87 @@ public virtual void Process()
             EmitIsGroundedChangedEvent(isGrounded);
         }
 
+        /// <summary>
+        /// Solves body collisions by not moving the body in case it can't go to its current position.
+        /// </summary>
+        /// <remarks>
+        /// If body collisions should be prevented this method needs to be called right before or right after applying any form of movement to the body.
+        /// </remarks>
+        public virtual void SolveBodyCollisions()
```
Comment (bddckr) [inline Prefabs/Locomotion/BodyRepresentation/SharedResources/Scripts/BodyRepresentationInternalSetup.cs]: This was kinda on the public facade on purpose - see the remarks section to understand why a user may want to have direct access to this.

One could argue that it's more internal and the user just has to understand that once they have a reason to call this themselves they are forced to work with these internals in a way after all.

Comment (bddckr) [reply Prefabs/Locomotion/BodyRepresentation/SharedResources/Scripts/BodyRepresentationInternalSetup.cs]: I'm dumb. It's still available on the facade, just proxied into this InternalSetup!

**Pull request merged**